consensus/consensus_v2.go

```diff
@@ -762,8 +765,43 @@ func (consensus *Consensus) onCommit(msg *msg_pb.Message) {
 		Int64("numReceivedSoFar", consensus.Decider.SignersCount(quorum.Commit)).
 		Logger()
 	logger.Info().Msg("[OnCommit] Received new commit message")
+
+	// proceed only when the message is not received before
+	if signed := consensus.Decider.ReadBallot(
+		quorum.Commit, validatorPubKey,
+	); signed != nil {
+		logger.Debug().
+			Msg("[OnCommit] Already received commit message from the validator")
+		// If signer already signed, and the block height is the same, then we
+		// need to verify the block hash, and if block hash is different, then
+		// that is a clear case of double signing
+		if h1, h2 :=
+			signed.BlockHash,
+			recvMsg.BlockHash; signed.BlockHeight == recvMsg.BlockNum &&
+			bytes.Compare(h1.Bytes(), h2.Bytes()) != 0 {
+			go func() {
+				// TODO(Edgar) Compute the aggregate vote, bitmap
+				consensus.SlashChan <- slash.Record{
+					BlockHash:   h2,
```

Comment:
Better add both block hashes and the signatures. So this record is verifiable by itself. Otherwise, it has to come with the crossLink to prove the double signing.

Reply:
gotca, will add the double signed block hash as well